simulator/main.go

```diff
@@ -9,60 +9,138 @@ package main
-// socat -d -d -d -d pty,link=/tmp/master,raw,echo=0 pty,link=/tmp/slave,raw,echo=0
 func main() {
 	c := make(chan os.Signal, 1)
 	signal.Notify(c, os.Interrupt, syscall.SIGTERM)
 
-	protocol = flag.String("protocol", ProtocolTCP, "Modbus protocol")
-	port = flag.String("port", ModbusPort, "Modbus port")
+	protocol := flag.String("protocol", protocolTCP, "Modbus protocol")
+	flag.Parse()
 
-	server := mbserver.NewServer()
-	defer func() {
-		server.Close()
-		fmt.Printf("Simulator shutdown. \n")
-	}()
+	_, scalabilityTestMode := os.LookupEnv(startingPortEnvName)
 
-	if *protocol == ProtocolTCP {
-		url := fmt.Sprintf("0.0.0.0:%s", *port)
-		fmt.Printf("Modbus TCP address: %v \n", url)
-
-		if err := server.ListenTCP(url); err != nil {
-			fmt.Printf("Failed to start the Modbus TCP server, %v\n", err)
+	if scalabilityTestMode {
+		err := createScalabilityTestSimulators()
+		if err != nil {
+			log.Fatalf("Fail to create devices for scalability test, %v. \n", err)
 		}
+
+		defer func() {
+			for _, device := range devices {
+				device.Close()
+			}
+			log.Printf("Close %d mock devices. \n", len(devices))
+		}()
+
 	} else {
-		config := &serial.Config{
-			Address:  "/tmp/master",
-			BaudRate: 19200,
-			DataBits: 8,
-			StopBits: 1,
-			Parity:   "N",
+		if *protocol == protocolTCP {
+			err := createTCPDevice(defaultDevicePort)
+			if err != nil {
+				log.Fatalf("Fail to create TCP device, %v. \n", err)
+			}
+		} else {
+			err := createRTUDevice()
+			if err != nil {
+				log.Fatalf("Fail to create RTU device, %v. \n", err)
+			}
 		}
-		fmt.Printf("Modbus RTU config %v \n", config)
 
-		if err := server.ListenRTU(config); err != nil {
-			fmt.Printf("Failed to start the Modbus RTU server, %v\n", err)
-		}
 	}
 
-	fmt.Printf("Start up a Modbus %s simulator. \n", *protocol)
 	<-c
+	log.Println("Modbus simulator shutdown.")
+}
+
+func createTCPDevice(port int) error {
+	device := mbserver.NewServer()
+	url := fmt.Sprintf("%s:%d", host, port)
+	if err := device.ListenTCP(url); err != nil {
+		return err
+	}
+	devices = append(devices, device)
+	log.Printf("Start up a Modbus mock device with address %s \n", url)
+	return nil
+}
+
+func createRTUDevice() error {
+	device := mbserver.NewServer()
+	config := &serial.Config{
+		Address:  "/tmp/master",
+		BaudRate: 19200,
+		DataBits: 8,
+		StopBits: 1,
+		Parity:   "N",
+	}
+	fmt.Printf("Modbus RTU config %v \n", config)
+	if err := device.ListenRTU(config); err != nil {
+		return err
+	}
+	devices = append(devices, device)
+	log.Printf("Start up a Modbus mock device with address %v \n", config)
+	return nil
+}
+
+func scaleDevice(startingPort int, simulatorNumber int) (scaledDevicePorts []int, err error) {
+	log.Printf("Create simulator, startingPort is %d, simulatorNumber is %d \n", startingPort, simulatorNumber)
+	count := 0
+	for count < simulatorNumber {
+		err := createTCPDevice(startingPort)
+		if err != nil {
+			return nil, err
```

Comment:
print out the error message.  you should show which port is unavailable

Reply:
The error print at https://github.com/edgexfoundry/device-modbus-go/pull/162/files#diff-a78ce53529228c405845d25a31ffcc9fR142